### ISSUE TITLE
Resource with caching enabled + sort + distinct filter breaks

### ISSUE THREAD

**Issue by jaydorsey** [JSONAPI-Resources/jsonapi-resources#1272]
## This issue is a (choose one):

tldr; You can't use `distinct` on a filter/recordset, plus a sort order, with caching enabled, when using postgres

- [x] Problem/bug report.
- [ ] Feature request.
- [ ] Request for support. **Note: Please try to avoid submitting issues for support requests. Use [Gitter](https://gitter.im/cerebris/jsonapi-resources) instead.** 

## Checklist before submitting:

- [x] I've searched for an existing issue.
- [x] I've asked my question on [Gitter](https://gitter.im/cerebris/jsonapi-resources) and have not received a satisfactory answer.
- [x] I've included a complete [bug report template](https://github.com/cerebris/jsonapi-resources/blob/master/lib/bug_report_templates/rails_5_master.rb). This step helps us and allows us to see the bug without trying to reproduce the problem from your description. It helps you because you will frequently detect if it's a problem specific to your project.
- [x] The feature I'm asking for is compliant with the [JSON:API](http://jsonapi.org/) spec.

## Description

Choose one section below and delete the other:

### Bug reports:

Setup:

- Rails 5.2.3
- Ruby 2.6.3
- jsonapi-resources 0.9.5 (also tested in 0.10.x)
- postgres 11.x
- redis for caching

We have an index endpoint that's hitting a resource with `caching` enabled, using a query parameter similar to this:

    ?sort=name&filter[name]=john

I receive an error from postgres below:

```
E, [2019-07-26T13:46:14.916989 #62203] ERROR -- : Internal Server Error: PG::InvalidColumnReference: ERROR:  for SELECT DISTINCT, ORDER BY expressions must appear in select list
LINE 1: ...WHERE (lower(contacts.name) LIKE 'john') ORDER BY "contacts"...
```

- A similar error occurs with Postgres (peeps-uuid fork example linked below)
- This occurs in both 0.9.5 & 0.10.x
- This doesn't happen with sqlite3. i know postgres & sqlite3 (and mysql) handle certain things differently, but I don't think this is so much a database issue as it is an issue with how cache-keys are calculated 
  - I do get a different error with sqlite3 (wrong number of args w/ mget)
  - Using a memory store instead of redis makes this error go away too (haven't tested outside of this script however)
- Removing caching, everything works fine (so the query itself is OK)

The root cause, I believe, is in [this line](https://github.com/cerebris/jsonapi-resources/blob/v0.9.5/lib/jsonapi/resource.rb#L1097) because we're only plucking the ID & cached field, but we also need the name (because of the sort/order passed in)

Bug report template:

https://gist.github.com/jaydorsey/b6f290fa2f1b7a002b6e734033373ff7

Fork/branch of peeps-uuid showing the same issue with Postgres:

https://github.com/jaydorsey/peeps-uuids/pull/1

**Comment by jaydorsey** [comment]
Closing this for now because I see an issue with my template that I need to fix (may not be an issue with sqlite)

**Comment by lgebhardt** [comment]
As you say the solution for this is to pluck any additional fields used in sorting and filters, but that's going to be difficult to get right since those fields could be introduced in a callable and may not even be on the primary table.

Could you explain why are you calling `distinct` in this case?

**Comment by jaydorsey** [comment]
> Could you explain why are you calling `distinct` in this case?

My example app doesn't need the distinct, it's purely to illustrate, but the actual code I was using did need it due to the joins (and the type of filter we wanted to run)

If I remember correctly, we have an object with a has_one :client, but also a has_many :contacts (same table) where the contact could appear multiple times and this particular filter (for better or for worse, probably the latter) is meant to search across both relationships and returning single instances of the records where they are a contact/client.

I agree this is going to be difficult, so if you want to close this ticket as wontfix that would be OK. We disabled caching for the time being because of this issue and another which I haven't been able to pinpoint quite yet. If I get more time, I'd like to take a look at both this and caching overall because we did see some noticeable improvements in certain endpoints/resources

**Comment by lgebhardt** [comment]
That's pretty much what I assumed you were doing, and it seems like a perfectly valid use case to me. I'm going to leave this open as I'd like to come up with a solution.

I haven't tested it, but I think in v0.10 I think you could leave off the distinct and the ruby code will (inefficiently) get you the right result. Might be an ok option if the number of related records is smallish.

**Comment by jaydorsey** [comment]
> That's pretty much what I assumed you were doing, and it seems like a perfectly valid use case to me. I'm going to leave this open as I'd like to come up with a solution.

That sounds good to me
 
> I haven't tested it, but I think in v0.10 I think you could leave off the distinct and the ruby code will (inefficiently) get you the right result. Might be an ok option if the number of related records is smallish.

When I tested 0.10 I only checked to see if this behavior was there, not if it de-duped the records for me. However, I think you might be right because I remember seeing some odd behavior with the number of records returned (I expected duplicates, but couldn't get them) that I attributed to an issue with the example I built rather than a change in gem behavior

I'll take another look at doing the upgrade to v10 so I can re-enable caching. In our case, the cache improvement on the other endpoints is worth a slight inefficiency on this one endpoint (that we are discussing refactoring anyways)
